js-watcher/cache_collection.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"context"
+	"log"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type CacheCollection map[string]*Cache 
+
+func (c *CacheCollection) synchronize(ctx context.Context) {
+	doneCh := make(chan bool)
+	for namespace, cache := range *c {
+		log.Printf("Starting sync for %s\n", namespace)
+		go cache.synchronize(ctx, doneCh)
+	}
+	syncCount := 0
+	log.Println("Waiting on cache synchronization")
+	for {
+		cacheSyncOK := <-doneCh
+		if !cacheSyncOK {
+			log.Fatalf("Failed to sync cache\n")
+		}
+		syncCount += 1
+		log.Printf("Synced %d/%d caches\n", syncCount, len(*c))
+		if syncCount == len(*c) {
+			break
+		}
+	}
+	log.Println("Synced all caches!")
+}
+
+func (c *CacheCollection) run(ctx context.Context) {
+	for	namespace, cache := range *c {
+		log.Printf("Starting cache for %s\n", namespace)
+		go cache.run(ctx)
+	}
+}
+
+func (c *CacheCollection) getAll() ([]runtime.Object, error) {
+	output := []runtime.Object{}
+	for _, cache := range *c {
+		res, err := cache.getAll()
+		if err != nil {
+			return nil, err
+		}
+		output = append(output, res...)
+	}
+	return output, nil
+}
+
+func (c *CacheCollection) getByUserId(userId string) ([]runtime.Object, error) {
+	output := []runtime.Object{}
+	for _, cache := range *c {
+		res, err := cache.getByUserId(userId)
+		if err != nil {
+			return nil, err
+		}
+		output = append(output, res...)
+	}
+	return output, nil
+}
+
+func (c *CacheCollection) getByNameAndUserId(name string, userId string) ([]runtime.Object, error) {
+	output := []runtime.Object{}
+	for _, cache := range *c {
+		res, err := cache.getByNameAndUserId(name, userId)
+		if err != nil {
+			continue
+		}
+		if res != nil {
+			output = append(output, res)
+		}
+	}
+	return output, nil
+}
+
+func (c *CacheCollection) getByName(name string) ([]runtime.Object, error) {
+	output := []runtime.Object{}
+	for _, cache := range *c {
+		res, err := cache.getByName(name)
+		if err != nil {
+			continue
```

Comment:
This looks v strange; usually if `err != nil` we need to handle it.

Reply:
I never return 404 on looking for a resource that does not exist. I simply return an empty list if you specify a server name that is not currently present. Luckily `k8s.io/apimachinery/pkg/api/errors` provides `IsNotFound` which checks if a k8s error is triggered by the resource missing. So I will `continue` only in this case.